### DISCUSSION TITLE
Automatic pipelines

### DISCUSSION THREAD

**Discussion by steckdenis** [SpinalHDL/SpinalHDL#764] [Ideas]
Dear all,

This idea is inspired from some difficulties I have understanding VexRiscv (and NaxRiscv seems to be similar). Basically, the VexRiscv plugins plug stuff into the pipeline stages of the processor, and the code is full of "was this generated in that stage?" conditions. The pipelining code checks that no Stageable is read before it is produced, but it seems difficult to really manage things, and the different plugins need to be aware of each other (or at least know when a Stageable may become available).

So, I was wondering if it would be possible to design a pipelining library in which the plugins receive inputs (Stageables for instance) and produce outputs. Then, LatencyAnalysis is used to compute the latency between the inputs and outputs. Finally, the inputs and outputs of the plugins (note here "plugins" not "stages") are wired together, with register stages added automatically so that, in the case of a processor, all the Stageables of a single instruction are always together in the pipeline. It may even be possible for all these inputs and outputs to be transparently wrapped into Streams so that we don't have to play with haltByOther and so on.

It would allow to write plugins like (very simplified and inexact code):

```
val io = new Bundle {
 val src1 = in Bits(32 bits)
 val src2 = in Bits(32 bits)
 val alu_op = in Something
 val register_writeback = out Bits(32 bits)
}

when (isFiring() && io.alu_op == MY_ALU_OP) {
  io.register_writeback := io.src1 + io.src2
}
```

Bonus: maybe it would even be possible to add a configuration parameter to Pipeline so that extra stages are automatically inserted to relax timing, at the cost of branch penalties, without any changes to the code.

The big advantage I would personally see is to be able to write a simple plugin that requires an extra pipeline stage (a real "fetch" stage that is not in VexRiscv) without having to modify VexRiscv, Pipeline and all the other plugins that assume that "decode" is the first stage.

**Comment by kleinai** [comment]
I've been casually interested in building out a full featured pipeline library. "Casually" as I don't have direct need of advanced features but it's an interesting problem.

There's a few different concepts I've been trying to capture in a pipeline library. This includes things such as automatic pipelining, stage latency analysis (goes along with pipelining), clean (or at least easy) streams in and out. There's other more advanced topics I think would be neat to explore such as anti-tokens for pipeline early evaluation or proper loop handling. I think all of these could coexist given in the right framework. Some of these features are just applied graph theory.

Defining an API which works well in Scala and feels nice to use has been a struggle.

**Reply by steckdenis** [reply]
At some point, I thought that the easiest and most transparent (readable) would simply to have pipeline plugins be Components with standard io. Each io would be a Stream, so that the plugin can access the .fire, .ready and .valid attributes of them. I count on the fact that a smart synthesizer will see that all these inputs are streams that are arbitrated the same, and "merges" that to something that actually looks like one big Stream of all the inputs (/outputs) of the plugin.

I mention "plugin" above because I expect some plugins to contain register slices (FPU for instance), so a plugin may span over several pipeline stages. I would also not assume that all the inputs of a plugin come from the same stage, so that a plugin can do stuff between decode and execute, and something else between writeback and fetch (a jump or a fault for instance).

**Comment by Dolu1990** [comment]
Here is a example of the dev pipeline api : 
https://github.com/SpinalHDL/SpinalHDL/commit/fdf6f6ac93e118757d504cc08332eaa3d6b45caf#diff-05c403f8f66c85c7c3c288ce516a465473f1c38827ad099f8c045fe1d4568806R1152
so, stage definition is manual, i think if we want to go into the more hardware abstracted ways, we would need to look at "data-flow" / "hls" paradigms ?
